### PR TITLE
retries...

### PR DESCRIPTION
**Pull request by TingluoHuang**

### CODE REVIEW

File: src/Agent.Worker/Build/BuildServer.cs [hunk 1]
```diff
@@ -9,28 +9,63 @@
...
+        private VssConnection _connection;
+        private Build2.BuildHttpClient _buildHttpClient;
 
-        public BuildServer(VssConnection connection, Guid projectId)
+        public async Task ConnectAsync(VssConnection connection)
         {
-            ArgUtil.NotNull(connection, nameof(connection));
-            ArgUtil.NotEmpty(projectId, nameof(projectId));
+            _connection = connection;
+            int attemptCount = 5;
```
Comment (stephenmichaelf) [inline src/Agent.Worker/Build/BuildServer.cs]: attemptsRemaining?

File: src/Agent.Worker/Build/BuildServer.cs [hunk 2]
```diff
@@ -9,28 +9,63 @@
...
+                {
+                    Trace.Info($"Catch exception during connect. {attemptCount} attempt left.");
+                    Trace.Error(ex);
+                }
+
+                await Task.Delay(100);
+            }
+
+            _buildHttpClient = _connection.GetClient<Build2.BuildHttpClient>();
+            _hasConnection = true;
```
Comment (stephenmichaelf) [inline src/Agent.Worker/Build/BuildServer.cs]: Can we hit this without having a connection? If _connection.HasAuthenticated stays false and we run out of attempts we would hit this line and not have an authenticated connection. We may need to check HasAuthenticated again and set _hasConnection based on that? Unless I am misunderstanding the use of _hasConnection :)

EDIT: It's somewhat subtle that ConnectAsync will throw on failure. Can we add a few comments? Or maybe change the response type to boolean? If we are trying to connect it doesn't seem exceptional for the connection to fail.

File: src/Agent.Worker/Build/FileContainerServer.cs [hunk 3]
```diff
@@ -14,47 +14,84 @@
...
+                {
+                    Trace.Info($"Catch exception during connect. {attemptCount} attempt left.");
+                    Trace.Error(ex);
+                }
+
+                await Task.Delay(100);
+            }
+
+            _fileContainerHttpClient = _fileContainerClientConnection.GetClient<FileContainerHttpClient>();
+            _hasConnection = true;
```
Comment (stephenmichaelf) [inline src/Agent.Worker/Build/FileContainerServer.cs]: Same comments as above.

File: src/Agent.Worker/Build/BuildServer.cs [hunk 4]
```diff
@@ -9,28 +9,63 @@
...
+    }
+
+    public sealed class BuildServer : AgentService, IBuildServer
+    {
+        private bool _hasConnection;
+        private VssConnection _connection;
+        private Build2.BuildHttpClient _buildHttpClient;
 
-        public BuildServer(VssConnection connection, Guid projectId)
+        public async Task ConnectAsync(VssConnection connection)
```
Comment (ericsciple) [inline src/Agent.Worker/Build/BuildServer.cs]: Why roll our own retry? Ted claimed the http client retries if settings are passed correctly. Are we passing the settings wrong?

Comment (bryanmacfarlane) [reply src/Agent.Worker/Build/BuildServer.cs]: Agreed.  What are we doing with this PR?

**Pull request closed without merging**